modules/services/window-managers/bspwm/default.nix

```diff
@@ -5,70 +5,73 @@ with lib;
+    builtins.replaceStrings upperChars (map (c: "_${c}") lowerChars);
 
-  formatConfig = n: v:
+  formatMonitor = monitor: desktops:
+    "bspc monitor ${strings.escapeShellArg monitor} -d ${
+      strings.escapeShellArgs desktops
+    }";
+
+  formatSetting = n: v:
     let
-      formatList = x:
-        if isList x then
-          throw "can not convert 2-dimensional lists to bspwm format"
-        else
-          formatValue x;
-
-      formatValue = v:
-        if isBool v then
-          (if v then "true" else "false")
-        else if isList v then
-          concatMapStringsSep ", " formatList v
-        else if isString v then
-          "${lib.strings.escapeShellArg v}"
-        else
-          toString v;
-    in "bspc config ${n} ${formatValue v}";
-
-  formatMonitors = n: v: "bspc monitor ${n} -d ${concatStringsSep " " v}";
-
-  formatRules = target: directiveOptions:
+      vStr = if isBool v then
+        boolToString v
+      else if isInt v || isFloat v then
+        toString v
+      else if isString v then
+        strings.escapeShellArg v
+      else
+        throw "unsupported setting type for ${n}";
+    in "bspc config ${strings.escapeShellArg n} ${vStr}";
+
+  formatRule = target: directives:
     let
       formatDirective = n: v:
-        if isBool v then
-          (if v then "${camelToSnake n}=on" else "${camelToSnake n}=off")
-        else if (n == "desktop" || n == "node") then
-          "${camelToSnake n}='${v}'"
-        else
-          "${camelToSnake n}=${lib.strings.escapeShellArg v}";
-
-      directives =
-        filterAttrs (n: v: v != null && !(lib.strings.hasPrefix "_" n))
-        directiveOptions;
-      directivesStr = builtins.concatStringsSep " "
-        (mapAttrsToList formatDirective directives);
-    in "bspc rule -a ${target} ${directivesStr}";
-
-  formatStartupPrograms = map (s: "${s} &");
+        let
+          vStr = if isBool v then
+            if v then "on" else "off"
+          else if isInt v || isFloat v then
+            toString v
+          else if isString v then
+            v
+          else
+            throw "unsupported rule attribute type for ${n}";
+        in "${camelToSnake n}=${vStr}";
+
+      directivesStr = strings.escapeShellArgs (mapAttrsToList formatDirective
+        (filterAttrs (n: v: v != null) directives));
+    in "bspc rule -a ${strings.escapeShellArg target} ${directivesStr}";
+
+  formatStartupProgram = s: "${s} &";
 
 in {
-  options = import ./options.nix {
-    inherit pkgs;
-    inherit lib;
-  };
+  meta.maintainers = [ maintainers.ncfavier ];
+
+  options = import ./options.nix { inherit pkgs lib; };
 
   config = mkIf cfg.enable {
-    home.packages = [ bspwm ];
-    xsession.windowManager.command = let
-      configFile = pkgs.writeShellScript "bspwmrc" (concatStringsSep "\n"
-        ((mapAttrsToList formatMonitors cfg.monitors)
-          ++ (mapAttrsToList formatConfig cfg.settings)
-          ++ (mapAttrsToList formatRules cfg.rules) ++ [''
```

Comment:
I think this was used to remove trailing whitespace since some people dislike having empty lines in their configuration file.

I'm impartial to the change.